src/gluonts/dataset/tabular.py

```diff
@@ -0,0 +1,226 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+from typing import Iterator, List, Union
+from datetime import datetime
+
+import pandas as pd
+
+from gluonts.dataset.common import Dataset, DataEntry, ProcessDataEntry
+from gluonts.dataset.field_names import FieldName
+
+
+class DataFrameDataset(Dataset):
+    """
+    A pd.DataFrame-based dataset type.
+
+    This dataset uses the long format for each variable. Target time series values,
+    for example, are stacked on top of each other rather than side-by-side. The
+    same is true for other dynamic or categorical features. An example of a dataframe
+    is given below.
+
+
+    Parameters
+    ----------
+    data: pd.DataFrame
+        pd.DataFrame containing at least timestamp, target and item_id columns.
+    target: str or List[str]
+        Name of the column that contains the target time series.
+        For multivariate targets `target` is expecting a list of column names.
+    timestamp: str
+        Name of the column that contains the timestamp information.
+    item_id: str
+        Name of the column that, when grouped by, gives the different time series.
+    freq: str
+        Frequency of observations in the time series. Must be a valid pandas frequency.
+    feat_dynamic_real: List[str]
+        List of column names that contain dynamic real features.
+    feat_dynamic_cat: List[str]
+        List of column names that contain dynamic categorical features.
+    feat_static_real: List[str]
+        List of column names that contain static real features.
+    feat_static_cat: List[str]
+        List of column names that contain static categorical features.
+    past_feat_dynamic_real: List[str]
+        List of column names that contain dynamic real features only for the history.
+    prediction_length: int
+        For target and past dynamic features last `prediction_length` elements are
+        removed when iterating over the data set. This becomes important when the
+        predictor is called.
+    """
+
+    def __init__(
+        self,
+        data: pd.DataFrame,
+        target: str,
+        timestamp: str,
+        item_id: str,
+        freq: str,
+        feat_dynamic_real: List[str] = None,
+        feat_dynamic_cat: List[str] = None,
+        feat_static_real: List[str] = None,
+        feat_static_cat: List[str] = None,
+        past_feat_dynamic_real: List[str] = None,
+        prediction_length: int = None,
+    ) -> None:
+        self.data = data.copy()
```

Comment:
Is there an actual need to copy the data?

Reply:
No, we only read from `data`. I removed the copy.